schemas/typedspans/README.md

```diff
@@ -0,0 +1,278 @@
+
+# Typed Spans
+
+## Rationale
+
+Spans represent different _canonical_ _types _of operations. Examples are
+
+* Local operations like method invocations
+* HTTP requests (inbound and outbound)
+* Database requests
+* eneric RPC requests like GRPC
+
+A span consists of a number of mandatory and optional attributes that add information
+about the represented operation to it.
+
+Depending on the _canonical type_ of an operation some attributes might be needed
+to represent and analyze a span correctly.
+
+**Example:** A HTTP request needs a status code to distinguish successful or
+unsuccessful operations.
+
+Right now, spans can be created freely and it’s up to the implementor to set all
+the attributes needed to represent the given type.
+
+## Existing work
+
+* There is already a _Type_ field that describes if a span is a parent or child span
+* There is already a _SpanKind_ field that describes if a span is a server or
+  client span - this distinguishes inbound and outbound requests.
+* There should be already a TypedSpanBuilder for Java
+
+## Proposal
+
+* Implement a **TypedSpanBuilder** for all platforms that ensures that all
+  needed attributes and the type is set.
+* Add a field **CanonicalType** that contains the type of span.
+
+### Proposed types
+
+(derived from [this doc](https://docs.google.com/spreadsheets/d/1H0S0BROOgX7zndWF_WL8jb9IW1PN7j3IeryekhX5sKU/edit#gid=0) by @discostu)
+
+The following canonical types should be implemented.
+The attributes are taken from the OC spec first with a fallback to the OT spec if
+there is no equivalent in OC. Additionally, some attributes that may enrich the
+span further are proposed.
+
+**Bold** = mandatory.
+
+#### HttpClient
+
+<table>
+  <tr>
+   <td><strong>CanonicalType = HTTP</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>SpanKind = CLIENT</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>http.method</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>http.host</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>http.path</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>http.status_code
```

Comment:
Feels like this should be required, but is only optional because it is unknown at span creation, which hints at a design issue. For example, if we used typed events instead, there could be HTTP_END event with status_code as required attribute.

Reply:
In case of a timeout, there's also no status_code.

Reply:
No, it's optional here because it was in the original doc.
Right now, OpenCensus fills the attributes in the end. The proposed addition is agnostic to when they are filled.
The proposal contains an optional check of required attributes on end but I am not sure if this is too rigid.